core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleWriter.scala

```diff
@@ -106,6 +108,19 @@ private[spark] class HashShuffleWriter[K, V](
       writer.commitAndClose()
       writer.fileSegment().length
     }
+    // rename all shuffle files to final paths
+    shuffle.writers.zip(sizes).foreach { case (writer: DiskBlockObjectWriter, size: Long) =>
+      if (size > 0) {
+        val output = blockManager.diskBlockManager.getFile(writer.blockId)
+        if (output.exists()) {
+          writer.file.delete()
+        } else {
+          if (!writer.file.renameTo(output)) {
+            throw new IOException(s"fail to rename ${writer.file} to $output")
```

Comment:
I think these partitions are independent, they should be OK whenever it's generated in different attempt, or that's the basic idea of how RDD works (could be re-run and got the same result).